### PR TITLE
[Kernel] Integrate CUTLASS MoE kernel with PPLX

### PR DESCRIPTION
**Pull request by ElizaWszola**
Integrate CUTLASS MoE fp8 kernels with PPLX.

Unit tests:
```
tests/kernels/moe/test_pplx_cutlass_moe.py
```

E2E testing:
```
export MASTER_ADDR=127.0.0.1
export MASTER_PORT=29500
export VLLM_ALL2ALL_BACKEND=pplx
python3 examples/offline_inference/data_parallel.py \
        --model="nm-testing/DeepSeek-Coder-V2-Lite-Instruct-FP8" \
        --dp-size=2 \
        --tp-size=1 \
        --trust-remote-code
```

### CODE REVIEW

File: tests/kernels/moe/test_pplx_cutlass_moe.py [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+    has_pplx = False
+
+from torch.multiprocessing import (
+    spawn)  # pyright: ignore[reportPrivateImportUsage]
+from typing_extensions import Concatenate, ParamSpec
+
+NUM_EXPERTS = [40, 64]
+TOP_KS = [6, 8]
+
+P = ParamSpec("P")
```
Comment (bnellnm) [inline tests/kernels/moe/test_pplx_cutlass_moe.py:36]: We should probably put all these multiprocess utilities in a separate file now since they are also used by test_pplx_moe.py

File: tests/kernels/moe/test_pplx_cutlass_moe.py [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+@pytest.mark.parametrize("e", NUM_EXPERTS)
+@pytest.mark.parametrize("topk", TOP_KS)
+@pytest.mark.parametrize("per_act_token", [True, False])
+@pytest.mark.parametrize("per_out_ch", [True, False])
+@pytest.mark.parametrize("world_dp_size", [[2, 1]])  #, [4, 2]])
+@pytest.mark.skipif(
+    (lambda x: x is None or not ops.cutlass_group_gemm_supported(x.to_int()))(
+        current_platform.get_device_capability()),
+    reason="Grouped gemm is not supported on this GPU type.")
+def test_cutlass_moe_pptx(
```
Comment (bnellnm) [inline tests/kernels/moe/test_pplx_cutlass_moe.py:301]: Type pptx -> pplx

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 3]
```diff
@@ -812,6 +847,7 @@ def __init__(
         assert quant_method is not None
         assert isinstance(quant_method, FusedMoEMethodBase)
         self.quant_method = quant_method
+        self.quant_method.moe = moe
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/layer.py:850]: This seems a bit sketchy to me.  If the `quant_method` needs a `MoEConfig` it should be part of the constructor or passed around as an argument.

File: vllm/model_executor/layers/fused_moe/pplx_prepare_finalize.py [hunk 4]
```diff
@@ -110,9 +121,11 @@ def prepare(
             out_expert_x_scale=expert_x_scale,
             dp_x=a1q,
             dp_x_scale=a1q_scale,
-            indices=rank_topk_ids,
+            indices=rank_topk_ids.to(torch.uint32),
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/pplx_prepare_finalize.py:124]: You shouldn't need this cast anymore. The type of the topk_ids can be controlled by passing torch.uint32 via indices_type to select_experts.

File: vllm/model_executor/layers/fused_moe/pplx_prepare_finalize.py [hunk 5]
```diff
@@ -110,9 +121,11 @@ def prepare(
...
             bound_m=bound_m,
         )
+        if expert_x_scale is not None:
+            expert_x_scale = expert_x_scale[:, :, 0:1]
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/pplx_prepare_finalize.py:128]: Is this the same as `expert_x_scale.view(-1, -1, 1)`?

File: vllm/model_executor/layers/fused_moe/pplx_prepare_finalize.py [hunk 6]
```diff
@@ -58,13 +60,22 @@ def prepare(
...
+                                                       self.quant_dtype,
+                                                       self.per_act_token,
+                                                       self.block_shape)
+        else:
+            repeat_rows = a1.shape[0]
+            a1q, a1q_scale = moe_kernel_quantize_input(a1, a1_scale,
+                                                       self.quant_dtype,
+                                                       self.per_act_token,
+                                                       self.block_shape)
+
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/pplx_prepare_finalize.py:76]: nit: can you collapse these branches together?  e.g.
```
repeat_rows = 1 if self.per_act_token else a1.shape[0]
a1q, a1q_scale = moe_kernel_quantize_input(
                a1,
                a1_scale if not self.per_act_token else None,
                self.quant_dtype,
                self.per_act_token,
                self.block_shape)
```
or even simpler if `a1_scale` is None iff `self.per_act_token` you can just pass `a1_scale` directly.

File: csrc/quantization/cutlass_w8a8/moe/moe_data.cu [hunk 7]
```diff
@@ -101,3 +101,35 @@ void get_cutlass_moe_mm_data_caller(
...
 }
+
+__global__ void compute_pplx_data(
+    int32_t* expert_offsets, int32_t* problem_sizes1, int32_t* problem_sizes2,
+    const int32_t* __restrict__ expert_num_tokens,
+    const int64_t* __restrict__ non_zero_expert_idxs, const int padded_m,
+    const int n, const int k) {
+  int expert_idx_out = threadIdx.x;
+  int expert_idx_in = non_zero_expert_idxs[expert_idx_out];
+  expert_offsets[expert_idx_out] = expert_idx_in * padded_m;
```
Comment (tlrmchlsmth) [inline csrc/quantization/cutlass_w8a8/moe/moe_data.cu:112]:
```suggestion
  int expert_idx_in = static_cast<int32_t>(non_zero_expert_idxs[expert_idx_out]);
  expert_offsets[expert_idx_out] = expert_idx_in * padded_m;
```

Could you doublecheck and try to not add any warnings to the build? (The implicit down-conversion here looks safe enough to me, but best to avoid implicit conversions)

File: csrc/quantization/cutlass_w8a8/scaled_mm_entry.cu [hunk 8]
```diff
@@ -242,6 +250,28 @@ void get_cutlass_moe_mm_data(
...
+void get_cutlass_pplx_moe_mm_data(
+    torch::Tensor& expert_offsets, torch::Tensor& problem_sizes1,
+    torch::Tensor& problem_sizes2, const torch::Tensor& expert_num_tokens,
+    const torch::Tensor& non_zero_expert_idxs, const int64_t num_local_experts,
+    const int64_t padded_m, const int64_t n, const int64_t k) {
+  // This function currently gets compiled only if we have a valid cutlass moe
+  // mm to run it for.
+  int32_t version_num = get_sm_version_num();
+#if (defined ENABLE_CUTLASS_MOE_SM90 && ENABLE_CUTLASS_MOE_SM90) || \
+    (defined ENABLE_SCALED_MM_SM100 && ENABLE_SCALED_MM_SM90)
```
Comment (tlrmchlsmth) [inline csrc/quantization/cutlass_w8a8/scaled_mm_entry.cu:262]: Should this be the following?
```suggestion
#if (defined ENABLE_CUTLASS_MOE_SM90 && ENABLE_CUTLASS_MOE_SM90)) \
```

I'm not sure why we're looking at `ENABLE_SCALED_MM_SM90`, but the check for `ENABLE_SCALED_MM_SM100` definitely looks wrong

File: tests/kernels/moe/test_pplx_cutlass_moe.py [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+from vllm.model_executor.layers.fused_moe.cutlass_moe import CutlassExpertsFp8
+from vllm.model_executor.layers.fused_moe.fused_moe import fused_topk
+from vllm.model_executor.layers.fused_moe.modular_kernel import (
+    FusedMoEModularKernel)
+from vllm.model_executor.layers.fused_moe.pplx_prepare_finalize import (
+    PplxPrepareAndFinalize)
+from vllm.platforms import current_platform
+
+try:
+    from pplx_kernels import AllToAll  # or AllToAllInternode?
```
Comment (tlrmchlsmth) [inline tests/kernels/moe/test_pplx_cutlass_moe.py:21]: IIUC, `AllToAll` dispatches to `AllToAllInternode` under the hood, so we shouldn't need to interact with it directly

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 10]
```diff
@@ -558,6 +543,21 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
...
+    def select_gemm_impl(self, prepare_finalize):
+        from vllm.model_executor.layers.fused_moe.cutlass_moe import (
+            CutlassExpertsFp8)
+
+        # self.moe should have been set by the caller
+        assert self.moe is not None
+
+        experts = CutlassExpertsFp8(
+            ((self.moe.num_experts + prepare_finalize.world_size - 1) //
+             prepare_finalize.world_size), self.moe.in_dtype,
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py:555]: I think it would be nice to factor out `(self.moe.num_experts + prepare_finalize.world_size - 1) // prepare_finalize.world_size` into a local variable for clarity, as it would help explain what it's doing.

**Review by tlrmchlsmth — COMMENTED** [review]
Left a few comments, but looks good overall -- lets try to get it landed once those and Bill's comments are addressed!

File: vllm/model_executor/layers/fused_moe/modular_kernel.py [hunk 11]
```diff
@@ -322,9 +323,17 @@ def forward(
 
         output = a1 if inplace else torch.zeros_like(a1)
 
+        # compute padded_M for PplxPrepareAndFinalize
+        if (hasattr(self.prepare_finalize, 'max_num_tokens')
+                and hasattr(self.prepare_finalize, 'world_size')):
+            padded_M = (self.prepare_finalize.max_num_tokens *
+                        self.prepare_finalize.world_size)
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/modular_kernel.py:330]: Should this be `max_num_tokens * (world_size // dp_size)`?

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py [hunk 12]
```diff
@@ -112,86 +108,142 @@ def apply(
             local_topk_ids = topk_ids
 
         topk = local_topk_ids.shape[1]
+        local_E = w1.shape[0]
+
+        if expert_num_tokens is not None:
+            non_zero_mask = expert_num_tokens[:] != 0
+            masked_local_E = int(non_zero_mask.sum().item())
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/cutlass_moe.py:115]: Is this going to interfere with cudagraphs?

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py [hunk 13]
```diff
@@ -112,86 +108,142 @@ def apply(
...
+            non_zero_expert_idxs = torch.nonzero(non_zero_mask).flatten()
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/cutlass_moe.py:116]: I think `nonzero` might cause trouble also.

File: vllm/model_executor/layers/fused_moe/pplx_prepare_finalize.py (hunk 5 above)
Comment (ElizaWszola) [reply vllm/model_executor/layers/fused_moe/pplx_prepare_finalize.py:128]: No, this is taking only one slice from `expert_x_scale`'s last dim. this is related to the scale format required by `dispatch()`.

File: csrc/quantization/cutlass_w8a8/moe/moe_data.cu (hunk 7 above)
Comment (ElizaWszola) [reply csrc/quantization/cutlass_w8a8/moe/moe_data.cu:112]: I don't see this producing any new warnings, I'll make the change

File: vllm/model_executor/layers/fused_moe/modular_kernel.py [hunk 14]
```diff
@@ -322,9 +323,17 @@ def forward(
...
+        else:
+            padded_M = M
+
```
[the remark this answers is not in the document]
Comment (ElizaWszola) [reply vllm/model_executor/layers/fused_moe/modular_kernel.py:333]: I'll go for the second suggestion

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py (hunk 12 above)
Comment (ElizaWszola) [reply vllm/model_executor/layers/fused_moe/cutlass_moe.py:115]: potentially... I think I can circumvent it with a custom CUDA kernel and extra mapping for `expert_offsets` and `problem_sizes1/2` if needed

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py [hunk 15]
```diff
@@ -206,67 +258,80 @@ def cutlass_moe_fp8(
...
+    workspace13 = torch.zeros(ws1, device=a.device, dtype=out_dtype)
+    workspace2 = torch.zeros(ws2, device=a.device, dtype=out_dtype)
+
+    if apply_router_weight_on_input:
+        assert topk_ids.shape[
+            1] == 1, "topk_ids must be 1 for apply_router_weight_on_input"
+        a = a * topk_weights.to(a.dtype)
+
+    from vllm.model_executor.layers.fused_moe.utils import _fp8_quantize
+    a1q, a1q_scale = _fp8_quantize(a, a1_scale, per_act_token)
```
[the remark this answers is not in the document]
Comment (ElizaWszola) [reply vllm/model_executor/layers/fused_moe/cutlass_moe.py:308]: This should be the function that runs the old version of cutlass MoE when no `PrepareAndFinalize` is being run. I changed the structure of functions/classes in this file, is it less messy now?

[the remark this answers is not in the document]
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/cutlass_moe.py:327]: I was thinking more in terms of not duplicating code and using the new modular classes to serve as the implementation of `cutlass_moe_fp8`.   `MoEPrepareAndFinalizeNoEP` should be able to do all the preparation/finalization and doesn't do any communication.

File: vllm/model_executor/layers/fused_moe/pplx_prepare_finalize.py (hunk 5 above)
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/pplx_prepare_finalize.py:120]: Can you elaborate on that?  I copied the setup of `expert_x_scales` directly from pplx's `test_all_to_all.py` test so I assumed it would have the proper format already.

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 16]
```diff
@@ -262,38 +281,45 @@ def init_prepare_finalize(self, moe: MoEConfig,
...
+                    0 if moe.quant_dtype.itemsize != 1 else
+                    ((moe.hidden_dim + moe.block_size - 1) // moe.block_size *
+                     torch.float32.itemsize)),
                 group_name=all2all_manager.cpu_group.group_name,
             )
 
             handle = all2all_manager.get_handle(all_to_all_args)
 
+            input_activations = get_quant_config_input_activations(
+                quant_config)
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/layer.py:298]: Is this guaranteed to always exist?  I've tried running `Qwen/Qwen3-30B-A3B-FP8` and it didn't appear to have an `input_activations` field in its `quant_config`.

I think this is probably fine for now but maybe with a `None` check.  There's a more general problem of finding the proper quantization info for each MoE layer that needs to be solved.

Comment (ElizaWszola) [reply vllm/model_executor/layers/fused_moe/layer.py:337]: This function returns `None` when `input_activations` is not available. There's also a `None` check everywhere where its result is used (or should be if I'm not missing anything)

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (ElizaWszola) [reply vllm/model_executor/layers/fused_moe/cutlass_moe.py]: I've updated it, can you have another look?

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 16 above)
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/layer.py:337]: Ok, maybe I was looking at a previous version of the PR.

**Review by bnellnm — APPROVED** [review]
LGTM

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py [hunk 17]
```diff
@@ -13,110 +13,94 @@
...
+    assert w1.shape[0] == w2_scale.shape[0], "w2 scales expert number mismatch"
+    assert a2_scale is None or a2_scale.dim(
+    ) == 0 or a2_scale.shape[0] == 1 or a2_scale.shape[0] == a1q.shape[
+        0], "Intermediate scale shape mismatch"
+    assert out_dtype in [torch.half, torch.bfloat16], "Invalid output dtype"
+    if expert_map is not None:
+        assert expert_num_tokens is None
+
+    M = a1q.shape[0]  # no pplx
+    padded_M = a1q.shape[1]  # pplx
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/fused_moe/cutlass_moe.py:80]: IIUC we have two cases:
* No PPLX
  * a1q is `total_num_tokens` by `hidden_size`
  * `expert_num_tokens` is not `None`
* PPLX
  * a1q is `num_experts` by `max_num_tokens_per_expert` by `hidden_size`
  * `expert_num_tokens` is `None`

And then we mostly differentiate between the two cases by checking if `expert_num_tokens` is `None`. IMO this is pretty confusing. Let's try to make it less confusing -- Could you add a comment explaining the situation? And then I think we should add a bool called `use_pplx = expert_num_tokens is None`, which would help the code be more self-documenting.

Comment (ElizaWszola) [reply vllm/model_executor/layers/fused_moe/cutlass_moe.py:80]: I've updated it, does it look better?

**Review by tlrmchlsmth — APPROVED** [review]

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py (hunk 17 above)
Comment (tlrmchlsmth) [reply vllm/model_executor/layers/fused_moe/cutlass_moe.py:80]: Looks great, thank you!

**Comment by bennorris123** [comment]
Hi @ElizaWszola - Is the pplx kernel already in the vllm image by default now? Or do we still need to go through the additional installation steps of expert parallelism (https://github.com/vllm-project/vllm/tree/main/tools/ep_kernels). Thanks!
